BLSToExecutionChange message generation error reporting and tests

infosecual opened the pull request:
This PR adds more granular error reporting and increased test coverage for the `validator credential set` command:

- Functional changes
    - `beacon/chaininfo.go`
        - `FetchValidatorInfo` will now report a more granular error when a provided public key is the wrong length
    -  `cmd/validator/credentials/set/process.go`
        - `generateOperationFromMnemonicAndPath`
            - check/err no `validatorKeyPath` provided
            - check/err if no validators were found even when there is no error returned from `generateOperationFromSeedAndPath` (not bubbling up errors here is valid functionality)
        - `generateOperationFromMnemonicAndValidator`
            - check/err for no validator found in chain
        - `generateOperationsFromMnemonic`
            - check/err for no validator found in chain
        - `generateOperationsFromPrivateKey`
            - check/error for private key 0x prefix
            - check/error for invalid private key length
            - check/error for validator not found
        - `parseWithdrawalAddress`
            - check/error that a withdrawal address was provided
            - check/error that withdrawal address is formatted correctly
- Testing changes
    - Increases test coverage in `cmd/validator/credentials/set/process.go` from 26.7% to 41.3%
    - Added 28 tests to the `validatorcredentialsset` package
        - New `TestGenerateOperationsFromPrivateKey` with 8 tests
        - New `TestGenerateOperationsFromMnemonic` with 4 tests
        - Added 7 more tests to `TestGenerateOperationFromMnemonicAndPath`
        - Added 8 more tests to `TestGenerateOperationFromMnemonicAndValidator`
        - Added 1 test to `TestGenerateOperationFromSeedAndPath`
   
NOTE: these remaining functions still need tests. I plan to add them in another PR in the coming days:

- `generateOperationsFromMnemonicAndPrivateKey`
- `generateOperationsFromAccountAndWithdrawalAccount`
- `generateOperationsFromAccountAndPrivateKey`
- `generateOperationsFromValidatorAndPrivateKey`

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -160,14 +160,28 @@ func (c *command) generateOperationFromMnemonicAndPath(ctx context.Context) erro
 
 	validatorKeyPath := c.path
+	// err if no path is provided
+	if validatorKeyPath == "" {
```
This function is only called if `c.path` is non-nil, so this check is not necessary.

infosecual replied:
Done. I also removed the `TestGenerateOperationFromMnemonicAndPath/NoPathProvided` I added since it is not needed.

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -192,7 +206,7 @@ func (c *command) generateOperationFromMnemonicAndValidator(ctx context.Context)
 				fmt.Fprintf(os.Stderr, "Gone %d indices without finding the validator, not scanning any further\n", maxDistance)
 			}
-			break
+			return fmt.Errorf("failed to find validator using the provided mnemonic, index=%s, pubkey=%#x", c.validator, validatorInfo.Pubkey)
```
The validator input could be something other than an index.

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -160,14 +160,28 @@ func (c *command) generateOperationFromMnemonicAndPath(ctx context.Context) erro
+	// validator was not found in our case (it should be found if a path is
+	// provided) and return an error if not.
+	if !found {
+		return errors.New("No validator found with the provided path and mnemonic, please run with --debug to see more information")
```
Errors should not be capitalized; this will fail linting.

infosecual replied:
Done. I tried to find other instances where I did this.

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -238,10 +252,14 @@ func (c *command) generateOperationsFromMnemonic(ctx context.Context) error {
 	maxDistance := 1024
 	// Start scanning the validator keys.
 	lastFoundIndex := 0
+	found_validator_count := 0
```
Please use camel case for variables; this will fail linting.

infosecual replied:
Done.

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -238,10 +252,14 @@ func (c *command) generateOperationsFromMnemonic(ctx context.Context) error {
 	lastFoundIndex := 0
+	found_validator_count := 0
 	for i := 0; ; i++ {
+		// if no validators have been found in the last 1024 indices, stop scanning
```
Comments need to start with capital letters and end with a period, or they will fail linting.

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -238,10 +252,14 @@ func (c *command) generateOperationsFromMnemonic(ctx context.Context) error {
-				fmt.Fprintf(os.Stderr, "Gone %d indices without finding a validator, not scanning any further\n", maxDistance)
+			// if no validators were found at all, return an error
+			if found_validator_count == 0 {
+
```
Unnecessary blank line.

infosecual replied:
Removed.

mcdee reviewed (changes requested):
Thank you for submitting this PR.  There are a few changes that should be made, and you should also run `golangci-lint run` against it to address some linting issues.

infosecual commented:
I addressed the comments and tried to find other places that would fail the linting cases you mentioned. I ran `golangci-lint run` (including with `--enable-all`) but it did not flag any of these cases. Any idea what I am missing here?

mcdee commented on cmd/validator/credentials/set/process.go:
```diff
@@ -192,7 +201,7 @@ func (c *command) generateOperationFromMnemonicAndValidator(ctx context.Context)
 				fmt.Fprintf(os.Stderr, "Gone %d indices without finding the validator, not scanning any further\n", maxDistance)
 			}
-			break
+			return fmt.Errorf("failed to find validator using the provided mnemonic, validator=%s, pubkey=%#x.", c.validator, validatorInfo.Pubkey)
```
The period needs to be removed here to satisfy the linter.

infosecual replied:
Done. I also fixed the test output that some of the linting changed.
